### ISSUE TITLE
Performance with many files

### ISSUE THREAD

**Issue by bernhardw** [jspm/jspm-cli#872]
Hello,

Thanks a lot for all the great work on jspm :thumbsup:

During development, when importing libraries split across many files such as React or Angular2, es6-module-loader loads all the required files individually which takes around 2-4 seconds. It helps to import a provided build instead, but not all packages include one.

Is there a suggested way around it?

**Comment by guybedford** [comment]
There are a number of approaches here, but no prescribed workflow currently.

In an ideal world the development server would be a fast HTTP/2 server with incremental caching understanding SystemJS and doing server-side transpilation and providing hot reloading.

Note that just using a better HTTP server for development can make a lot of difference. Some of them are very badly optimized for large numbers of requests.

@geelen is doing some amazing work on these sort of ideas in http://npmjs.org/package/jspm-server, but I don't know how ready it is.

While we wait for the future, the best workflow is simply to have common bundles that you load at the top of your app:

```
  jspm bundle app/**/* - [app/**/*] dependency-bundle.js
```

You then include `dependency-bundle.js` at the top of the page, and you'll only be loading your own application code as separate files in the browser.

Other users have quite advanced pre-compilation gulp tasks to do ES6 transpilation locally before loading separate files in the browser.

It's an area where users are still on there own for now, so please share what you learn with others as we collectively refine these workflows.

**Comment by bernhardw** [comment]
Thank you Guy for the detailed answer. I was hoping to get rid of my pre-compilation tasks during development, but creating a dependency bundle works for now.

**Comment by geelen** [comment]
[jspm-server](https://github.com/geelen/jspm-server) is definitely usable right now, and I'd love feedback on whether it improves your workflow. Basically, you can mark files as being "safe" to live-reload, and if those files change it won't have to reload the whole browser. It's still a work in progress, and will definitely benefit from changes coming down in the next version of JSPM & SystemJS, but right now I'm using it for all of my projects.

That said, doing the dependency bundling *will definitely help as well*, so that's a neat trick. I'd love there to be a higher-level command for doing that, something like `jspm bundle-deps` that bundles everything in `jspm_packages`? Would that be possible?

Also, @guybedford, you mention "better HTTP servers". Do you know of any? I want to find the best and build `jspm-server` around it. I don't know anything that supports HTTP/2 either. I have an issue for it here: https://github.com/geelen/jspm-server/issues/4

**Comment by OrKoN** [comment]
@geelen I am using nginx with spdy support here: https://github.com/OrKoN/jspm-ember-playground It's faster than any node-based servers I used. But I guess it would be hard to package nginx into the jspm-server.

Also there is node-spdy server https://github.com/indutny/node-spdy and https://github.com/molnarg/node-http2  which is not fully working right now (see https://github.com/molnarg/node-http2/issues/114).

**Comment by OrKoN** [comment]
I have also written a blog post about optimizing the performance with many files using gulp and nginx https://60devs.com/optimizing--default-jspm-workflow-with-gulp-and-nginx.html

@geelen I wonder if jspm-server can replace the gulp stuff I have. Is there a tutorial on how to migrate an app to jspm-server? Is there gulp-jspm-server or smth similar?

**Comment by geelen** [comment]
There's not really, jspm-server is sort of designed to be run independently of gulp or similar things.

I'd love to get a version of nginx statically linked and shipped inside an NPM package :smiley: I wonder if that's possible...

**Comment by OrKoN** [comment]
@geelen I think this should be possible similar to this https://jamie.curle.io/posts/webfaction-installing-custom-nginx/ But I guess it needs to be compiled on npm install.

**Comment by bernhardw** [comment]
I've set up a simple static file server based on [http2](https://github.com/molnarg/node-http2) to play around, but I see no difference in performance between http2 and http. It doesn't seem to multiplex. Any ideas why?

**Comment by jpray** [comment]
Regarding solutions other than HTTP/2... it seems that if you want even one unbundled ES6 module to be loaded while developing, then babel is what is affecting the load time because it dynamically loading dozens and dozens of dependencies.  @guybedford, is there a way we can bundle the modules for babel into a single file?

**Comment by OrKoN** [comment]
@bernhardw According to my tests there is no difference too. I guess http/2 may be helpful in production or if the server leverages the server push technologies. I have tried http2 module you mention but it didn't work with Chrome as a client. How did you get it work?

**Comment by bernhardw** [comment]
@OrKoN I also couldn't get [serve-static](https://www.npmjs.com/package/serve-static) to work, but it did work with the less used [node-static](https://www.npmjs.com/package/node-static).

```
var fs = require('fs');
var http2 = require('http2');
var fileServer = require('node-static');
var file = new fileServer.Server('./public');

var options = {
    key: fs.readFileSync('./localhost.key'),
    cert: fs.readFileSync('./localhost.crt')
};

var server = http2.createServer(options, function (req, res) {
    req.addListener('end', function () {
        file.serve(req, res);
    }).resume();
});

server.listen(8080);
```

**Comment by bernhardw** [comment]
@jpray Thanks, that did bring it down from ~1200ms to ~750ms. For anyone else trying to bundle the babel runtime, you can and should disable the runtime in your `config.js`:

```
System.config({
  "baseURL": "/",
  "transpiler": "babel",
  "babelOptions": {
    // "optional": [
    //   "runtime"
    // ],
    "blacklist": []
  },
  "paths": {
    "*": "*.js",
    "github:*": "jspm_packages/github/*.js",
    "npm:*": "jspm_packages/npm/*.js"
  }
});
```

That seems to be all there is to shave off. These are my network requests now:
![image](https://cloud.githubusercontent.com/assets/919880/8512642/2069619a-234d-11e5-970f-793a31cde728.png)

**Comment by guybedford** [comment]
Two perf tips along these lines, and why this is hard currently:
1. Finding a performant HTTP/2 server that does in-memory caching of files is hard. If you know any, please share :)
2. When serving `jspm_packages`, it should actually be served with a far-future expires header so that it is cached in the browser and never requested again. That way just the local application folders are refreshed. Again, knowing to set this and how to set it is not obvious when running a dev server.

**Comment by lookfirst** [comment]
The caching is really complicated though cause in dev mode I currently have to constantly flush the cache (using a nice chrome extension i found for single click expiration) because my `.map` files are heavily cached and out of date all the time. Any ideas on that one?

**Comment by guybedford** [comment]
@lookfirst only cache `jspm_packages`, not anything else (the code you write yourself). Because paths in jspm_packages are versioned, updates get forced. Linking workflows break down here, and need custom rules though.

**Comment by lookfirst** [comment]
@guybedford Right, but my point is that right now, I'm hard flushing the entire chrome cache every time I reload the browser in order to get .map files to work properly. That would kill any benefit of caching `jspm_packages`.

**Comment by guybedford** [comment]
I guess a no-cache header on the .map files?

**Comment by lookfirst** [comment]
Great idea! =) Lol... (feeling dumb, i should have thought of that)...

**Comment by lookfirst** [comment]
Sad face. https://github.com/BrowserSync/browser-sync/issues/719

**Comment by geelen** [comment]
@lookfirst have you looked into [jspm-server](https://github.com/geelen/jspm-server)? We should be able to add some caching rules in there (always cache jspm_packages, never cache `.map` files) pretty easily, no?

**Comment by lookfirst** [comment]
I'd rather use browsersync as the basis. I wish your server just extended
that. I also don't really see the value proposition of jspm-server over
what I've already got setup.

**Comment by geelen** [comment]
Well, it's aware of the dependencies within your SystemJS build, so it's able to live-reload modules within your application without having to reload the browser, which will always be quicker. But basically, I want a development server for JSPM that's as fast as possible out the box, without requiring any configuration (beyond adding some `export let __hotReload = true` statements around the place)

I had hoped I could port jspm-server to be an extension of BrowserSync but I think that might be too much work for not much gain at this point.

**Comment by lookfirst** [comment]
The gain is the functionality of browsersync without having to do much
work. If your code could live within the context of a middleware, that
would be optimal.

I'd worry the live reload of modules would break angular, no? It also seems
like there could be potential for stuff to get messy there. I'd hate to
have to debug a reload failure. Also adding __hotReload to all my files
isn't very ideal.

So far, what I've got setup is pretty darn easy and fast. Gulp manages the
watching of files and only recompiles ones (es6/less/templates) that
change. I use my gulp-helpers so I don't have to rewrite gulp files ever
time I start a new project.

https://github.com/lookfirst/gulp-helpers

Now, just waiting to see if browsersync can set the headers correctly in a
middleware and I'll be all set.

Thanks for piping in... I did look at jspm-server, but it just didn't
scratch my itch enough. =)

**Comment by geelen** [comment]
Yeah, angular makes it tough. I don't know exactly how to handle reloading there. And even with jspm-server, there's always times you have to fully reload, and the speed is still an issue.

@guybedford back to the original point. Could we do something silly like run SystemJS in a serviceworker? Have it manage transpilation & caching on the client side, not the server? It seems like a good fit to me, but maybe I'm missing something...

**Comment by guybedford** [comment]
@geelen my previous plan was to build a jspm server that did the plugin transpilation and es6 transpilation. Basically it would run the loader pipeline on the server first, and just send the ready sources to the client. Then you get browser-cached transpiled sources, and it's a solid foundation development server to use with the native loader too in future.

**Comment by lookfirst** [comment]
@guybedford I thought about that as well and in practice, it was just easier to setup gulp with browsersync. If running the loader pipeline on the server first means also getting a single file to the server, then that doesn't work well for me cause it makes it harder to debug things. The individual files are nice during development so that I can quickly see which file has issues and go to that in my IDE.

**Comment by geelen** [comment]
So the native loader won't do any transpilation?

That feels like a use-case that's already covered pretty well but the webpack dev server, and there are plenty of issues with it. Reloading (they call it Hot Module Replacement) is pretty complex when you have to serialise a chunk down to the browser, in particular. When the loader knows it's running in the browser (except in bundle mode of course) it can be a bit smarter, in my mind.

That's why I've thought in-browser caching would be the way forward. But you're right, a serviceworker is disconnected enough from the browser context that it'd be the same problems as I just mentioned. So maybe just adding some caching of transpiled sources into SystemJS (and in development only) is the simpler approach?

**Comment by guybedford** [comment]
@geelen the reason I liked the server though was that it would be equally useful in development and production. I do still plan to work on this over the coming months, but it will be a little while unfortunately. With browser caching we have a development-only optimization, that may be confused for a production workflow.

The above server is also the natural home for hot reloading, so would be nice to merge efforts. What about the serializing gets complex there? You have been working on this problem far more than me though, so interested to hear more from your experiences.

Also in terms of the future of hot-reloading, had an interesting chat with @paulirish, where he described a workflow of editing ES6 in the Chrome devtools and having that hot-reload. V8 optimizations can do a lot more there than any other approach (code updates without re-execution) so it seems like this is where things should go for hot-reloading eventually. This workflow would then (in some distant future where V8 supports modules) be based on the loader listening to the edit event of a transpiled module from devtools, and then sending back the transpilation to be hot-swapped without refreshing the page. Food for thought :)

**Comment by lookfirst** [comment]
Well, this should help... https://github.com/BrowserSync/browser-sync/issues/719

**Comment by guybedford** [comment]
:+1:

**Comment by lookfirst** [comment]
If you're following the above browsersync issue, I've submitted the PR's and they should get integrated tomorrow or the next day. Once those are in, I'll add the middleware to the gulp-helpers project. If you use that middleware, then proper browser caching of .map and jspm_packages files will be a great speed improvement without having to resort to crazy hot reload optimizations, etc... I like browsersync, it works great and is very very well tested... we should all focus on using that and not inventing more projects...

**Comment by Munter** [comment]
I have a strong feeling that all the tools I've been working on previously can fit into this space and make the development experience faster.

[Fusile](https://github.com/Munter/fusile) can set up a directory for you that mirrors your web root. When you read from it, the sources are transpiled, sourcemapped, autoprefixed, cached and watched. It exposes a file system as an API, so any static file server hooks right into it and gains all the benefits of static file caching.

[staticache](https://github.com/assetgraph/staticache) is an experiment in keeping an always updated dependency graph of your entire code base alive in an express middleware. It serves files from memory and with a far future cache expiry. By watching all files it can let file updates trigger cache bustin on only the relevant assets by trickling up md5 hashes in all urls that point to it. This will trickle up the dependency graph till it hits the root nodes. Initial load will be slow and load all files, but next reload will only load the entry point plus the path to any updated assets.

[expush](https://github.com/Munter/expush) is an express middleware that uses an in memory dependency graph of your static assets to do spdy push of the entire sub-graph you need for each request. This speeds up loads of many files in development both because of multiplexing, but also because a runtime module loader won't have to wait for the whole download, transpile, parse chain to finish before being able to incrementally discover the next dependency

**Comment by geelen** [comment]
This could be really really cool, I've been hoping we could do something together @munter :)

My vision basically involves the following:

- Keep the hosted JSPM because that's super incredible. Such a good way for me to, say, point to a plunkr like this (that has NPM & Github deps & es6 transpilation): http://plnkr.co/edit/ZyCLINzwZFusq2btH1pd?p=preview
- Keep the JSPM workflow as-is: all static files, no dev server required etc.
- Have a simple, no-config dev-server that simply speeds up your development experience.

I think the final thing should be *called* `jspm-server`, but I don't really mind how it's built or who maintains it. I just want it to have as nice an API as JSPM does :)

How does that sit with everyone?

**Comment by Munter** [comment]
I agree on those points. I want to augment the workflow on the premise of jspm and systemjs. The tools should be able to do the right thing based on configuration read out of package.json, config.js and possibly tool specific configuration.

Areas that are possible to improve on without having any impact on how you write code:

* Browser cache control
 * Automatic far future expire jspm_packages
 * Keep track of updates and dependency graph and somehow augment all sources with cache busting hashes in their outgoing relations. I'm unsure if/how SystemJS can work with that.
* http2/spdy for opening up more pipes through multiplexing
* http2/spdy push based on dependency graph knowledge
* Assuming css is loaded with `<link />` tags, css live reloading

Areas that are possible to improve, but might need some extra configuration:

* Handling transpiling on the server with caching. SystemJS possibly need to be instructed to skip most of its loading and transpiling pipeline because it's assumes it has already been take care of. Such a config.js augmentation could likely be automated in the server
* React hot loading

Anything I missed?

**Comment by lookfirst** [comment]
Ok, my changes are now in browsersync and I'm happy to report that WEEEE, things are much faster now and it is clear that the caching and expiration is working as expected. No more clearing the browser cache every time I change my .js files so that the map files work correctly!

Non-scientific speed tests: On a load of a single page with 131 requests (mostly from jspm_packages) and 2.6MB transferred, it takes 1.98s with a cleared cache. With a reload after the cache has been populated, it is 1.10s and 160KB transferred.

```
let expireHeaders = (proxyRes) => {
	if (proxyRes.req.path.indexOf('.map') !== -1 || proxyRes.req.path.indexOf('/js') !== -1) {
		proxyRes.headers['expires'] = 'Thu, 01 Dec 1983 20:00:00 GMT';
		proxyRes.headers['cache-control'] = 'public, max-age=0, no-cache, no-store';
	}
	if (proxyRes.req.path.indexOf('jspm_packages') !== -1) {
		proxyRes.headers['expires'] = 'Thu, 01 Dec 2021 20:00:00 GMT';
		proxyRes.headers['cache-control'] = 'private, max-age=3600';
	}
};
```

Use that function in your browsersync configuration. 

```
	proxy: {
		proxyRes: [expireHeaders]
	}
```

Turn on browser caching even when devtools is open so that the browser caches files correctly.

So while I'm sure there is more optimizations that can be had, this makes me really really happy.

**Comment by iksose** [comment]
Anyone know how he's accruing these metrics, and if there's plans to get jspm to competitive with webpack? I assume he's using bundle-sfx, since he's a core angular contributor. But I do know this includes a "micro-loaded implementation" which may be causing slow loads?

https://www.reddit.com/r/angularjs/comments/3da3fe/angular2_webpack_starter_updated_to_alpha31/ct85ouk

**Comment by guybedford** [comment]
Thanks, I've posted a reply at https://www.reddit.com/r/angularjs/comments/3da3fe/angular2_webpack_starter_updated_to_alpha31/cta2mgx.

**Comment by logiclabs** [comment]
For a HTTP/2 server, try Caddy: https://caddyserver.com/

**Comment by NervosaX** [comment]
@logiclabs It does not support push yet: https://github.com/mholt/caddy/issues/144

**Comment by f15gdsy** [comment]
@guybedford Hi, In terms of using "jspm bundle app/**/* - [app/**/*] dependency-bundle.js", 
I got this error when running it on terminal:
```
err  Expected operator before app/src/components
```

Any ideas?

**Comment by guybedford** [comment]
@f15gdsy try putting quotes around the `'app/**/* - [app/**/*]'` part for the terminal.

**Comment by f15gdsy** [comment]
@guybedford It works! Thanks! It reduces the compile time a lot!
But I'm wondering if there is a way to reduce it further? Like only compile what's changed in app/ and cache the unchanged ones?

**Comment by lookfirst** [comment]
@f15gdsy That sounds terrible. You should really only bundle for production. Your production should be built using a CI server. Doing that would be asking to do caching on your CI server which seems like a dangerous and brittle thing.

**Comment by f15gdsy** [comment]
@lookfirst Hi, I'm not sure what you mean. What I'm trying to do is to reduce the time to get my changes in code reflect to browser.
Here is what I'm doing in index.html
```html
<script src="./jspm_packages/system.js"></script>
<script src="./config.js"></script>
<script>
	System.import("./build/dependency-bundle.js");
	System.import("./app/src/components/App/App.js");
</script>
```
where dependency-bundle.js is all the dependencies of what's inside ./app, and App.js is the entrance of the application.

So with the help of 
```
jspm bundle app/**/* - [app/**/*] dependency-bundle.js
```
I don't need to compile the dependencies again when changing what's inside ./app.
But I still need to compile everything inside ./app.
So my question is is it possible to cache the unchanged code in ./app and only compile the changed ones to reduce compile time.

**Comment by guybedford** [comment]
@f15gdsy yes this completely a valid approach for development performance as the bundle doesn't need to be refreshed that regularly. Ideally such a bundle shouldn't be checked in to version control certainly though.

Better workflows around development with granular refreshing should exist certainly! If anyone is interested in working on these problems please do let me know. SystemJS builder does support build caching, so setting this up as a custom build task via `require('jspm').Builder` would be able to get the compile time down using the trace and compile caches (`builder.getCache`, `builder.setCache`). These features are still somewhat experimental so not currently documented, but let me know if you need further guidance here.

**Comment by lookfirst** [comment]
> "What I'm trying to do is to reduce the time to get my changes in code reflect to browser during development."

I write my code in ES6, so what works *great* for me is to run a gulp based build that sits and watches for changes and feeds the changed file(s) through babel. This change watching is smart and only recompiles *exactly* the files that change. So, I just made a change to `app.js` and this is the output:

```
[00:47:05] File hub/src/main/js/home/app.js was changed, running tasks: babel
[00:47:05] Starting 'babel'...
[BS] 1 file changed (app.js)
[00:47:06] Finished 'babel' after 754 ms
```
I have that integrated with browserSync [BS] which watches for changes on my compiled files so that I don't even have to reload my browser, it just reloads automatically. I use Intellij with auto save turned on which saves me the step of even saving the file.

Any bundling that I do with jspm happens as a production step only.

All of this setup with gulp is made trivial with my gulp-helpers project, it is literally one line in my gulpfile and since I use angular, it even handles the ngAnnotation for me...

```
taskMaker.defineTask('babel', {taskName: 'babel', src: path.source, dest: path.output, ngAnnotate: true, compilerOptions: babelOptions, watchTask: true, notify: true});
```

The `notify: true` is also awesome in that I get operating system level notifications when there are compile failures in babel.

Using the browserSync caching header code referenced above, the browser ends up getting back mostly 304 headers for each request. 

Sure, you can bundle all your dependencies into a separate file, but I just don't see any need to do that since I have the proper http caching headers turned on. I'm literally loading hundreds of files for each request and things happen quickly enough in development mode that it really isn't an issue.

**Comment by guybedford** [comment]
@lookfirst thanks for these points, sound like you have a nice workflow going! How are you refreshing SystemJS in the browser without reloading here, or is that specifically for assets? There's some effort to put some better dev docs together at https://github.com/jspm/jspm-cli/pull/1136, any feedback or assistance there is welcome.

**Comment by jackfranklin** [comment]
I just pulled out the very small module that we use at work into its own thing: https://github.com/jackfranklin/jspm-dev-builder.

It's a pretty straight forward setup that makes full use of the jspm file and trace caching to bundle our app up. We've seen pretty good perf on the cached bundles - down to 0.5 seconds or less, on our app which contains 800+ files.

This is an area that I'm really interested in so keen to see what directions people take.

**Comment by JonahBraun** [comment]
I did a bunch of testing with different servers. You can see the results: https://github.com/JonahBraun/jspm-perf-test

Unfortunately HTTP2 doesn't actually help (in fact, it hurts a bit). Just having Dev Tools open imposes a huge performance hit. I'd love to know why. When developing, it's normal to have Dev Tools open, but that huge of hit is unacceptable. Does anyone know why?

**Comment by OliverJAsh** [comment]
Jonah, did you have 'disable cache' enabled in DevTools?

**Comment by lookfirst** [comment]
@oliverjash is right...

**Comment by JonahBraun** [comment]
@OliverJAsh, @lookfirst, actually I'm purposefully ensuring a clean cache with a hard refresh.

My point is caching shouldn't be necessary, or at least shouldn't be a necessary first step. 700ms is acceptable for clean cache full build. Having Dev Tools balloon that to 4000ms without reason is crazy. I'd love to know why and how to stop that.

I also wanted to inform that HTTP2 doesn't provide a perf boost (at least on localhost). It's too bad browsers don't support H2 without TLS, as I think we would see something then. :(

**Comment by Munter** [comment]
Dev tools give you continuous feedback on what your updated situation is. This takes time. If you're on the dom tree and animate updates, this slows you down. If you're on the network tab while loading is going on this slows you down. If your on the source tab while loading things this slows you down. This is the reason, and it's not crazy.

Now, with 900+ files loaded in development in my work setup the SystemJS load is 2-3x slower than the corresponding requirejs setup. That part is a bit worrysome and points towards something either requiring a great deal more work on the client than before or there being a choke point somewhere.

**Comment by guybedford** [comment]
@Munter is there a speed up if you set all the module format information via metadata - `System.config({ meta: { 'app/*': { format: 'amd' } } })` for example? The automatic format detection may be providing some slowdown.

**Comment by JonahBraun** [comment]
@Munter, I'm not sure. I think it's an issue in DevTools. I'm not doing any DOM manipulation till the end. The extra latency seems excessive.

@guybedford Looking at the network tab, I noticed that there is very few parallel requests happening. I imagine that's because modules are being included as requested, and so walking the code base ends up being synchronized. But it's too bad, as making the bulk of the 160+ requests in parallel might solve the issue.

One suggestion: FB ships a single JS file for the React development version. It's not minified, and so is very usable for diagnosing issues. If this file was linked to by the short name in the JSPM registry, it would also solve the performance issue for React users and make the initial experience of JSPM users a lot better.

**Comment by capaj** [comment]
@JonahBraun you know I have my react app reloads always under 500ms even on slow machines with my https://github.com/capaj/jspm-hot-reloader
No need to do 160+ requests every time you change one file.

**Comment by jackfranklin** [comment]
> One suggestion: FB ships a single JS file for the React development version. It's not minified, and so is very usable for diagnosing issues. If this file was linked to by the short name in the JSPM registry, it would also solve the performance issue for React users and make the initial experience of JSPM users a lot better.

Was about to suggest that myself. Any reason why we couldn't do that? I see the motivation for loading all the bits of React individually but it would cut out a lot of requests in dev mode.

**Comment by capaj** [comment]
@jackfranklin we did something similar with socket.io-client https://github.com/jspm/jspm-cli/issues/780

**Comment by JonahBraun** [comment]
@capaj thanks for the link. Another great workaround. This issue itself should be fixed however. jspm should be able to at least match the performance of a compile-then-load tools. That is to say, JS in the browser is as performant as JS on the CLI, so performance should be similar.

I've done further research: the slow down is not caused by the XHR requests itself. It's something else SystemJS is doing when loading files or modules that is triggering the slowdown with DevTools.

JonahBraun/jspm-perf-test#1

**Comment by pietschy** [comment]
Thanks @jackfranklin, I just followed your approach and am seeing the bundle rebuild in under 300ms for single file changes.  Using browser sync in conjunction with Intellij autosave the browser updates in about a second.

**Comment by luisfmsouza** [comment]
#watching

**Comment by lookfirst** [comment]
Interesting single data point for those of you who think HTTP/2 is going to magically solve all the issues...

http://engineering.khanacademy.org/posts/js-packaging-http2.htm

```
Khan Academy spent several months rearchitecting its system to move from a package-based scheme to one that just served JavaScript source files directly, at least for clients that support HTTP/2.0. And here's what we found:

Performance got worse.

On investigation we found out there were two reasons for this:

We were serving more bytes due to reduced compression quality
The server had unexplained delays serving dozens of JS files
```

**Comment by capaj** [comment]
@lookfirst I would take that with a pinch of salt. Google app engine's webserver issues could be single source of slowdowns. And nobundling workflow's greatest strength isn't in a new user. It is in a usecase of as returning user. It is in caching.
HTTP2 enabled nonbundled app is cached much more efficiently, there is no argument around that.

**Comment by lookfirst** [comment]
@capaj **"Interesting single data point"**

**Comment by douglasduteil** [comment]
:+1:

**Comment by adiachenko** [comment]
I've played around with nginx 1.9.7 today and saw absolutely no benefit in enabling HTTP/2. In fact, it seems like page load is even slightly slower than usual (probably due to mandatory use of SSL encryption but I'm not sure).

**Comment by guybedford** [comment]
@adiachenko did you use depCache?

**Comment by adiachenko** [comment]
@guybedford No. It was my impression that depCache is intended for production use anyway.

For now I'm mostly interested in speeding up page load in development. I use vagrantbox with nginx because I'm back-end developer so it's quire convenient for me. By the way, enabling gzip and using native filesystem instead of virtualbox shared folders improved performance by 20%.

I'm wondering whether node based solutions would be a bit faster than nginx due to more "native" environment of execution. I already use browsersync for live-reloading scripts and injecting css on change but I haven't tried it yet (I need SPA routing and proxy pass to api so configuring it may be a bit troublesome). I'll share my results if I bring myself to test it.

**Comment by capaj** [comment]
@adiachenko I am under the impression that depCache should be usable in development as well as production. You should try it with depCache.

**Comment by guybedford** [comment]
Without depCache, page load time is restricted by `(latency + processing) * dependency` tree depth, so it is necessary for perf.

**Comment by adiachenko** [comment]
@capaj @guybedford it doesn't seem like depCache brings any difference for me although the app is mostly boilerplate at this point (a lot of dependencies, not so much actual code). Probably not the best target for benchmarking.

**Comment by peteruithoven** [comment]
Besides bundling my dependencies adding depcache didn't really seem to matter for me, only a few ms. The browser seems mostly busy with transpiling my es6 code. 
I'm quite happy with the jspm / systemjs hot reloader though, that really helps. I hope we can make that workflow slightly easier in the future, not having to add hot reload script in my html for example, having that managed through an init config in config.js (that's somehow automatically added) would be awesome.

**Comment by douglasduteil** [comment]
@peteruithoven I'm actually experimenting something similar with the same results...
https://github.com/geelen/jspm-server/issues/41

Checkout my repo https://github.com/douglasduteil/jspm-server
I published a fork on npm too https://www.npmjs.com/package/douglasduteil...jspm-server

**Comment by teohhanhui** [comment]
@capaj I tried using `capaj/systemjs-hot-reloader` but it doesn't work well with Docker at all (you have to use polling which kills performance). So that's too bad... (Also, I don't like the idea of hot reloading in the first place but that's just personal preference.)

@Munter: https://github.com/jspm/jspm-cli/issues/872#issuecomment-121925878
> Keep track of updates and dependency graph and somehow augment all sources with cache busting hashes in their outgoing relations. I'm unsure if/how SystemJS can work with that.

What about @guybedford's suggestion here?

> only cache jspm_packages, not anything else (the code you write yourself). Because paths in jspm_packages are versioned, updates get forced.

But there's a problem with that when we use the github registry with a branch name. Will changing it to use git commit hash (in directory name) solve the problem?

**Comment by egaga** [comment]
How to deal with the fact that one needs to refresh the whole application when a file changes? i.e. there's no easy-and-proper way to hot-reload angular application partially. So using hot-reloading (e.g. jspm-server) is not a solution.

Is the problem the following:  even if I change only one typescript file, I need to reload the page, and thus the jspm reloads using typescript-plugin all of the already-up-to-date files one-at-a-time which takes uncomfortably large amount of time.

**Comment by guybedford** [comment]
@egaga https://github.com/capaj/systemjs-hot-reloader is the project we want to use here.

**Comment by frederikschubert** [comment]
@egaga There is also the [angular2-hot-loader](https://github.com/mgechev/angular2-hot-loader/) that is currently being developed.

**Comment by mgamsjager** [comment]
@guybedford use as in make part of jspm or as preferred add-on?

**Comment by mikz** [comment]
I just made https://github.com/mikz/jspm-dev-server
It is HTTP/2 web server, that correctly sets far future expire cache headers for files in `jspm_packages`.
It is very raw, starts only on port 3000, servers only current folder, etc. But the source is trivial. https://github.com/mikz/jspm-dev-server/blob/master/index.js
I also want to add proxy support, to use with other webapps. 

It has https://github.com/capaj/chokidar-socket-emitter baked in, so works out of the box with https://github.com/capaj/systemjs-hot-reloader.

**Comment by ineentho** [comment]
Would it be possible to cache the babel/ts/traceur transpilation result in the browser (localStorage?) in order to speed up loading times? I haven't looked into it myself yet, but theoretically if the file was loaded from cache, you should be able to speed the process up significantly by using a precompiled version of the file, right?

**Comment by mikz** [comment]
IMO the main problem is the browser loading the files sequentially. Every file can require new dependency, so loading more other files etc. Babel transpile is negligible in this situation.

**Comment by ineentho** [comment]
@mkiz Are you sure about that? Even if I generate a dependency cache, I see 0-3 seconds used for downloading assets, and then a several second gap before anything is executed. I thought babel was the problem there, but I guess I'll have to do some more research

**Comment by phenomnomnominal** [comment]
I've seen the same thing as @ineentho, still digging into it.

**Comment by ineentho** [comment]
I've done some experimenting, but I don't have a big project to test on right now so I'm not sure how big of an effect it will have.

I cached the whole translate event into localStorage. If there has been no changes to an ES6 file, Systemjs doesn't even try to load babel, which quite a big save. The save on individual files isn't as big as I thought though, but I'll see tomorrow when I can test on a project at work.

If somebody want to check it out and see if you can improve anything, I created a gist: https://gist.github.com/ineentho/3ccaaec164e418f685d7

localStorage is quite limiting however with a hard limit of 5MB, so I think I will have to check out indexedDB or maybe even serviceworker cache somehow.

**Comment by mikz** [comment]
Very relevant: https://github.com/samccone/The-cost-of-transpiling-es2015-in-2016

**Comment by fyodorvi** [comment]
Having a project with 800+ files I tried HTTP/2 and changing caching, it doesn't improve anything.
I agree with @mizk, main problem seems to be the number of files itself and that browser loads them one by one, and add transpiling as well.
I think proper solution here would be using same approach as @jackfranklin does with his jspm-dev-builder, although I tried his module it didn't work, it just doesn't remove file from cache. 
But I like the general idea. You bundle everything into one file and memory-cache each of them. Once you have a file change - remove it from cache, re-bundle again and reload the browser. It should work perfectly and super-fast. That's the best solution.

**Comment by phenomnomnominal** [comment]
@fyodorvi Just curious, what are your start up times like with a production bundle?

**Comment by lookfirst** [comment]
According to DevTools in Chrome on a mac:

I have a fairly large app that in development mode makes 388 requests, 4.5m transferred and finished in 6.07s with a full cache flush. With a reload of the page, it is 386 requests, 693kb transferred and finished in 5.35s.

My production bundle is a single file, 770kb transferred and finishes in 2.70s with a cache flush. A reload is 14.7kb and 2.55s. This includes going over the network too (everything is hosted on appengine).

I'm using the same technique [I describe above](https://github.com/jspm/jspm-cli/issues/872#issuecomment-122990647). You can now see how my app has grown since then as well. 

=)

**Comment by fyodorvi** [comment]
@phenomnomnominal it's quite fast, about two seconds. I actually spent few more hours trying to figure out stuff with jspm-dev-builder... And it actually works! Load time is now dramatically short. It took about 25 seconds before to make a change and see it in the browser. Now it's 5 seconds. And if you reload the page - it's same two seconds as on prod. It feels like Christmas now. Source maps work as well, and BrowserSync. Probably I should make a demo project with it.

**Comment by peteruithoven** [comment]
@fyodorvi Please do. 
@jackfranklin @lookfirst I'm curious how you enable hot reloading in a workflow which reloads all javascript?
I'm curious whether the jspm-dev-builder could enable a quicker full reload, but then have systemjs-hot-reloader take over for smaller changes, enabling hot reloading.

**Comment by lookfirst** [comment]
I use angularjs so hot reloading is kind of a broken concept. =(

**Comment by nlwillia** [comment]
@ineentho your idea of a caching translate hook seems to work really well.  There may be more sophisticated solutions that will ultimately produce a better result, but this alone shaves a couple seconds off of a 4 second load for me.  `localStorage` is indeed pretty limiting, so I took a shot at using IndexedDB [here](https://gist.github.com/nlwillia/70303c2ccdb69ad3661d).

**Comment by peteruithoven** [comment]
@nlwillia That looks like a very interesting approach. And because you use Dexie there are fallbacks to web sql and localStorage. Any change you could release this as package? Using it should be a manner of importing it before any other module right?

**Comment by nlwillia** [comment]
I'd have to look at how loader plugins are structured for release, but yeah, probably.

For now, it's just:
```
<script src="jspm_packages/system.src.js"></script>
<script src="//npmcdn.com/dexie@1.3.3/dist/dexie.min.js"></script>
<script src="plugin-translate-cache.js"></script>
<script src="jspm.browser.js"></script>
<script src="jspm.config.js"></script>
<script>
	System.import(...);
</script>
```

**Comment by peteruithoven** [comment]
Have you tried adding `plugin-translate-cache.js` (which contains your gist i'm assuming) as the first import in the file you're importing with System.import()? This would require using some kind of import statement (cjs / esm) to import dexie.

**Comment by fyodorvi** [comment]
After doing couple of days of testing and experiments, I'm absolutely sure that there's no other reliable way of improving load speed that using build cache (jspm-dev-builder approach).

Here's a bit of statistics from my current project. 

Before using the approach it was ~800 dependencies loading for the project:

![before](http://i.imgur.com/xqOPoDs.png)

I happen to have quite powerful machine, but still it takes about 15 seconds to start up the project. Each time you reload the page browser serves you almost 800 files. For other developers on the project average time is 30 seconds. Some people running slow machines got 60 to 90 seconds (!). 

After applying the approach load time is production-like, and it should stay marginally the same even when project grows further:

![after](http://i.imgur.com/nXEXAjN.png)

Obviously, serving just one already transpiled file is way faster that loading 800 dependencies. But how about serving changes and rebundling? Well, that's possible thanks to Builder.invalidate method. Application has one entry point with all imports (it is Angular app). We are using gulp at the project, which has persistently running "watch" task. So gulpfile workflow is:
1. On startup bundle the application using 
```javascript
var builder = new require('jspm').Builder();
builder.bundle(input, output, options);
```
This may take a while (almost same time as loading the app without that approach in browser), takes ~20 seconds on my machine.
2. Run browserSync
3. Setup chokidar watcher to watch for changes in javascript and html files

Next, when file watcher sees file change we trigger builder.invalidate it from cache:
```javascript
devBuilder.invalidate(moduleName)
```
And then trigger `builder.bundle` again, which will happen in a second or two, because only 1 file needs to build, other are in cache already.

![build] (http://i.imgur.com/NxsJBvt.png)

Lastly, we trigger `browserSync.reload()`.

So using that approach I wait 4 to 5 seconds before I see the app reloaded in the browser. And while using the app if it's crashed (e.g. error or I just need to reload), it takes ~3 seconds to reload. Where's previously in was 15 seconds just to load the app in *any case* (not counting gulp tasks which applied as well).

I'll try to find some time to implement and publish proper boilerplate project for that approach. But you may do it yourself in the meantime.

I also apply this approach for unit tests (we're using karma+jasmine, but it's doesn't really matter).

**Comment by lookfirst** [comment]
@fyodorvi WOW, very interesting idea. I'll consider trying that soon.

**Comment by rubyboy** [comment]
@nlwillia that's a fantastic solution. I've used it in our project with some tweaks, to avoid loading as well as transpilation. Here's the modified version: https://gist.github.com/rubyboy/1722db5339ce546078e5

When using this approach it avoids loading all dependencies in the jspm_packages folder and only transpiles when needed.

Thanks for the tip!

**Comment by doctyper** [comment]
@nlwillia @rubyboy Wow, what a speed boost! Please do release this as a plugin, it will do wonders for using jspm in development.

**Comment by ineentho** [comment]
Which version of jspm are you using? I've found both the script I posted earlier and the forks unreliable under JSPM .16.

**Comment by fyodorvi** [comment]
@nlwillia @rubyboy what a nice solution, indeed! Best choice if you don't want to spend time configuring build process and invalidating cache. 
I get ~4 seconds loading time, still doesn't beat ~2.5 seconds bundled, but it's damn close. I wouldn't implement my solution if tried this before, it's good enough.

**Comment by nlwillia** [comment]
@ineentho I used 0.17 when working out the kinks.  The only issue I had to work around is that SystemJS apparently expects some basic initialization to be done, and skipping the default hook call circumvents that.  I tried it with 0.16 today, and it sort of worked, but not completely (one of my shim'd dependencies didn't load correctly and none of the CSS came in), so there may be more caveats there.

**Comment by fyodorvi** [comment]
@nlwillia @rubyboy 
I couldn't get your code properly running without that:
```javascript
...
if (!cached || cached.hash != file.hash) {
   db.files.where('url').equals(file.url).delete();
...
}
```
DB complained that key already existed on every change.

I fully tested your solution with file watcher to measure how long it takes to see changes in the browser running app with ~800 dependencies. And the result sits somewhere between 7 and 10 seconds. So my verdict is that it's better than not using any caching at all (and super easy), but it's slower than using sophisticated build process with SystemJS builder cache invalidation (which is way harder to configure), which gives you maximum of 5 seconds waiting for change and 3 seconds on page reload. Although, page reloading without needing to rebuild anything with your approach takes marginally the same time (about 4 seconds).
So guess using your client cache approach will suit most of the projects with smaller dependency tree.

**Comment by nlwillia** [comment]
Thanks for that.  I ran into the same problem a bit ago and have updated my gist.

My current project is smallish and reloads in ~4 seconds which still isn't great in my opinion, but it's a little less intolerable than it was.  Long term, browsers are either going to have to get a lot better at small files and loaders are going to have to get faster and smarter about transpilation or the things the browser is slow at are going to have to move elsewhere as you've been exploring.  Hot reloading is cool when it works, but it's not a universal solution, and caching is low hanging fruit but certainly not perfect.

**Comment by fyodorvi** [comment]
@nlwillia try pre-bundling approach that I've described, it should your reload time by 30-40%. My verdict is based on opinion that 5 seconds is acceptable waiting time, but if it's not good enough just try it. I'm actually keen to find out how it improves a smaller project.

**Comment by yyaari** [comment]
Hi, thank you all for sharing your efforts.
i tried using @nlwillia, @ineentho and @rubyboy  scripts. when changing a file and reloading i get a dexie error constraint stating the key already exists. and a 'Cannot read property 'lastIndexOf' of undefined' from SystemJs.

using jspm-dev-builder works as expected. about 4-5 seconds to build the bundle file with source maps when a file has changed (and much less if not using source maps)

**Comment by fyodorvi** [comment]
@yyaari, read my comment above to fix dexie issue.
For jspm-dev-builder try lowResSourceMaps: true build option, which makes source maps go per line, works much faster. Depending on your directory structure, you may also require to map your sources with chrome dev tools.

**Comment by yyaari** [comment]
@fyodorvi, lowResSourceMaps doesn't seem to work for me. when setting it to true, no source maps file is created.

**Comment by oligot** [comment]
Just to let you know, I've created a small app that uses jspm-dev-builder under the hood
https://github.com/oligot/jspm-dev-builder-cli
It's basically just a CLI app on top of the Node.js module.
Instructions on how to use it are on the Github page.

**Comment by fyodorvi** [comment]
@yyaari you need to have both sourceMaps: true and lowResSourceMaps: true.

**Comment by yyaari** [comment]
@fyodorvi thanks for the tip!  
I feel that the current flow i have of using jspm dev builder and auto reload with browsersync is preaty good although it's kind of missing the point of systemjs

**Comment by mikz** [comment]
I plan to try implement HTTP2 Server Push into the jspm-dev-server. That should eliminate the need for watch tasks and separate process. It could also precompile the whole bundle and pre push it to the browser without any temporary files or changing the html. Will see. See https://github.com/mikz/jspm-dev-server/issues/7 for progress.

**Comment by amcdnl** [comment]
@rubyboy - I dropped in your indexdb solution and had no impact on my results. I'm downloading a pretty heavy payload, running around 1042 requests on page load. Seeing about 19.89s for total finish.

**Comment by rubyboy** [comment]
@amcdnl what you should see is that the first time it loads all the files and on the second refresh it will use it straight from the cache. Notice it only does that for the jspm_packages folder. You can tweak the code to include anything else (update the match expression).
Maybe check on the resources tab in Chrome to confirm the cache was actually created.

**Comment by peteruithoven** [comment]
@amcdnl did you check the chrome profiler? This should help determining what is taking the most time. 
I was hoping someone would turn the idea into a package, then I'd love to check this out.

**Comment by amcdnl** [comment]
@rubyboy Yup, that number was with it cached and chrome showed them NOT coming down.

I wonder if the pure number files its fetching offset the perf ratio of using the index db solution? I did notice it faster when using localStorage but i quickly exceeded my storage limit. But that is also 'in memory' vs the indexdb being read disk.

**Comment by nlwillia** [comment]
Here's some followup on a few items:

1\.  I looked into packaging what I'd posted above as a [gist](https://gist.github.com/nlwillia/70303c2ccdb69ad3661d) as a plugin with a more conventional installation approach, but I don't see a good way of doing that within how SystemJS works.  Plugins are applied around package loader bindings usually based on a specific extension, and loader composition is tricky.  Everyone's personal usage scenario (and loader choices) may vary, but I'm interested in caching transpilation on everything that comes in, not just a specific source package.  Given that this is something of a [hack](https://github.com/systemjs/systemjs/pull/1158) to begin with and not especially hard to deploy one-off for the development scenario it pertains to, my recommendation is to just add these two lines to your html (right after `system.src.js`) and get on with your day.  I've been using the approach all week with pretty good results.
```
<script src="//npmcdn.com/dexie@1.3.3/dist/dexie.min.js"></script>
<script src="https://rawgit.com/nlwillia/70303c2ccdb69ad3661d/raw/5e57e0f7adc7838408a3351d50ad603fee6b2b93/plugin-translate-cache.js"></script>
```
Note that the first version I posted had a bug where it wasn't deleting the previous file before overwriting it.  Make sure to apply that fix if you have an older version of the gist or a derivative based on it.  Also note that there are some issues with 0.16 that I haven't gone back to try to figure out so YMMV.

2\.  It wasn't clear to me just what was involved with using [jspm-dev-builder](https://github.com/jackfranklin/jspm-dev-builder), but I had a chance to try it this afternoon, and it wasn't as complicated as I thought it would be.  This is basically a thin subclass of the [SystemJS builder](https://github.com/systemjs/builder/) that you feed events from a file watcher.  It initially creates a build of your whole project which takes a few seconds, but after that everything is in memory and it can usually rebuild the composite file in a couple of seconds.

I used @oligot's [example project](https://github.com/oligot/jspm-dev-builder-cli) to create a bare bones node process for my project.
```
npm install chokidar jspm-dev-builder --save-dev
```
```
// dev-builder.js
const DevBuilder = require('jspm-dev-builder');
const chokidar = require('chokidar');

var builder = new DevBuilder({
    expression: 'src/main',
    outLoc: 'dev-build.js',
    buildOptions: {
        sourceMaps: true,
        lowResSourceMaps: true
    }
});

builder.build();

chokidar.watch('**/*.ts').on('change', function(path) {
    builder.build(path);
});
```
```
node dev-builder
```
That's it.  Your needs may vary, but this does not take a lot of setup and isn't much more invasive than adding lines to your index file.  The only thing that my require some adjustment is whatever you're using as a [live-server](https://www.npmjs.com/package/live-server).  If it's watching for any change to reload the page then it will end up reloading twice because it will see the file you change and then it will see the build file updated by the builder.  In my case `live-server --watch=dev-build.js` took care of that.

One caveat is that `jspm-dev-server` depends on jspm 0.16.  My project is 0.17, and it initially errored because it could not find `config.js`.  I fixed this by editing `node_modules/jspm-dev-builder/package.json` to reference `"jspm": "^0.17.0-beta.11",` then did an `npm install` in that directory.  What it's doing does not look very version-specific, so it should work okay with 0.17.

Except that I then ran into [this issue](https://github.com/systemjs/builder/issues/529) after saving my first file change.  That's probably something new in the current version of SystemJS, and I was able to work around it by editing `node_modules/jspm-dev-builder/node_modules/systemjs/dist/system.src.js` and changing `mapped.substr` to `(mapped.node || mapped).substr` in the `validateMapping` function.  I didn't test 0.16; you might not run into the problem there.  With this hack, the builder then seemed to work as it should.

3\.  Performance.  This is on a small project.  I averaged 5 samples (ms).  I measured page load time by capturing a timestamp at the top of the body and subtracting elapsed time from it after my application had loaded.  Not super scientific, but consistent enough to give a feel for how these approaches compare at this scale.

| file change w/transpile cache | file change w/o transpile cache | file change w/builder (builder + reload = total) |
|-------------------------------|---------------------------------|--------------------------------------------------|
| 2242                          | 3215                            | 2190 + 862 = 3052                                |
| 92 requests, 47.3kb           | 90 requests, 6.5M               | 15 requests, 1.7MB                               |

As the number of files grows, the number of requests is going to start skewing things more towards the builder.  At work, I'm running nginx w/some aggressive cache rules to cut down on the bigger 3rd party dependency load times, but transpilation and bootstrapping ends up being where most of the time is spent.  In a project of this size, the numbers are not dramatically different, but this test is on a fast machine with a SSD.  Larger projects or slower machines are going to find increasing benefit from experimenting with one of these strategies.

**Comment by peteruithoven** [comment]
@nlwillia couldn't you do it as simple as my systemjs-debugger? Just overriding some methods. 
https://github.com/peteruithoven/systemjs-debugger/

**Comment by born2net** [comment]
I am hoping whatever the solution is, that jspm, the lib adopts it and adds it to the library as right now the biggest pain point is load times which can take up to 10 seconds in a large project... now x times you debug in an hour :/

**Comment by amcdnl** [comment]
@born2net my project takes around ~28s to load :S ...

**Comment by frederikschubert** [comment]
Is there any reason for not using [the hot reloader](https://github.com/capaj/systemjs-hot-reloader)? It reduces the reload time to ~10ms. I am using it in my company's angular 2 projects and it works great.

**Comment by amcdnl** [comment]
@frederikschubert nope! my app is ng1 though, ng1 + hmr is a hairy beast.

**Comment by born2net** [comment]
Ouch 28sec... I feel for you....

**Comment by born2net** [comment]
@frederikschubert  does it work with latest angular2 beta .11?

**Comment by frederikschubert** [comment]
Haven't tested it yet, but it should.

**Comment by born2net** [comment]
ok tx

**Comment by mgamsjager** [comment]
Care to share the project setup or ref to a gist?

**Comment by frederikschubert** [comment]
Its integrated in my company`s workflow but the hot reloading part is taken from the [example project for Angular 2](https://github.com/capaj/jspm-ng2).

**Comment by born2net** [comment]
would you say it's bug free? just sounds like magic... too good to be true, if you know what I mean...

**Comment by frederikschubert** [comment]
I didn`t notice any bug yet, but I am traveling right now and have not excessively tested it ;)

**Comment by amcdnl** [comment]
@born2net beta11 has a bunch of issues, stick w/ beta9 until new versions rolls out.

**Comment by amcdnl** [comment]
@born2net angular2 beta .11 has a lot of issues, stick w/ beta9 until beta12 comes out.

**Comment by born2net** [comment]
got it, tx!

**Comment by teohhanhui** [comment]
Hot reloading is another can of worms.

**Comment by oligot** [comment]
For info, I've just released a new version of _jspm-dev-builder-cli_
https://github.com/oligot/jspm-dev-builder-cli

It now uses the SystemJS builder _trace_ method to know which files to watch.
That way, we are sure to only watch the files that are really needed and not all the files of a given folder that have nothing to do with jspm (like a README.md file for example).

Another change is the use of [watchman](https://facebook.github.io/watchman/) (via the Node.js [sane](https://www.npmjs.com/package/sane) module) instead of chokidar.
chokidar turns out to be quite buggy for me (at least in Vim).
Changed files were not detected at all, or detected as removed instead of changed.
watchman seems to be more reliable (at least in my early tests), so we'll see if it turns out to be a good solution...

Feedback is welcome :smirk:

**Comment by amcdnl** [comment]
@oligot Why not just use gulp.watch?

**Comment by oligot** [comment]
Gulp is more that a file watcher: it's a task runner.
I find it easier to use a module that does just one thing (watch for files changes) than another one that among other things can watch files.

**Comment by amcdnl** [comment]
@oligot absolutely understand, was just curious.

**Comment by guybedford** [comment]
@oligot great work. I'd love to merge this into jspm core actually. I hope you don't mind if I copy the approach you've taken? If you'd like to help with the process just let me know too.

**Comment by guybedford** [comment]
(tracking in https://github.com/jspm/jspm-cli/issues/1572)

**Comment by oligot** [comment]
Feel free to copy the approach. I can also help to merge it into jspm core.

**Comment by fyodorvi** [comment]
Published [JSPM Watch](https://www.npmjs.com/package/jspm-watch) module - my version of file watcher that manages JSPM bundling and cache invalidation.

Features:

- proper dependencies addition/removal handling: ideally, no need to restart the process ever
- minimum and straightforward configuration
- hassle-free file watching with built-in chokidar
- both windows and unix support
- unit tests build support
- proper compilation error handling
- informative console output
- non-javascript modules support - css and html

```
npm i jspm-watch
```

Minimum config:

```javascript
var JspmWatch = require('jspm-watch');

watcher = new JspmWatch({
   app: {
      watch: ['src/**/*.js', 'src/**/*.html', 'src/**/*.css'],
      input: 'src/my-app.js',
      output: 'dist/my-app.js'
   },
   jspm: require('jspm')
});

watcher.start();
```

Full documentation as well as demo project: (https://github.com/fyodorvi/jspm-watch).
I would really appreciate your comments and suggestions.

**Comment by guybedford** [comment]
jspm 0.17-0-beta.12 now supports `jspm bundle x --watch` and `jspm build x --watch` commands. Many thanks to @jackfranklin, @AdamBurgess, @oligot and @fyodorvi for helping push this work forward.

**Comment by egaga** [comment]
@guybedford After a quick test with 0.17 beta, reloading after changes is as fast as it should be, at least for a small app. This was a real pain, so thank you!

**Comment by jackfranklin** [comment]
@guybedford so happy about this and pleased I was able to help along the way. Awesome stuff!

**Comment by born2net** [comment]
where can I find docs on how to upg to 0.17 from 0.16? 
as things don't work for me (for example CSS directory error)

**Comment by amcdnl** [comment]
@guybedford any docs out for this?

**Comment by OrKoN** [comment]
@amcdnl @born2net http://jspm.io/0.17-beta-guide/development-bundling.html

**Comment by zakdances** [comment]
Does this not support TypeScript?

> jspm bundle ./src/app.ts -wid



```
Building the bundle tree for ./src/app.ts...

err  Error on fetch for app/app.ts.js at file:///Users/zak/Zak/projects/MyProject/src/app.ts.js
  Error: ENOENT: no such file or directory, open '/Users/zak/Zak/projects/MyProject/src/app.ts.js'
    at Error (native)
```

**Comment by born2net** [comment]
I had the same issue with CSS and gave up on this, hoping maybe it become more stable soon... for now I can't use this build with TypeScript :( ( I get CSS error on ng2 so not sure if same exact reasoning )

**Comment by zakdances** [comment]
Yeah, I really don't understand this. Running a command like `jspm bundle ./src/app.ts -wid` doesn't stop the dev server from making a billion requests...it just does the same thing it always (requesting and loading each file individually) does PLUS it injects `build.js` bundle, which is then full of redundant code.

Is the 

```
<script>
        SystemJS.import('app.ts');
    </script>
```

supposed to be removed or edited in some way? Is production mode supposed to be switched on from `config.js`?

**Comment by mgamsjager** [comment]
Are you on 0.17-beta12 ?

**Comment by zakdances** [comment]
@mgamsjager Yes. And I'm using it with `lite-server 2.2.0`

**Comment by mgamsjager** [comment]
And when not using the short parameter but --watch --inject ? 

Disabled cache?
See build.js being loaded?

**Comment by zakdances** [comment]
@mgamsjager I tried it both ways. I checked the sources tab in Chrome Developer Tools, and both my unbundled code AND the `build.js` bundle are being loaded. I've also tried putting `jspm` into "[production mode](http://jspm.io/0.17-beta-guide/bundling.html)" (although I don't know what that is because it's not explained in the docs). No difference.

`lite-server` just won't stop making requests for every single file. :(

**Comment by zakdances** [comment]
Apparently switching `production mode` to `false` just replaces all the `304` requests with `200` requests. The requests are still made regardless. (I've also tried disabling the cache in Developer Tools) This is my network trace when I run my server after `jspm bundle ./src/app.ts -wid`:

![image](https://cloud.githubusercontent.com/assets/1066961/14408060/442e8630-feaa-11e5-94c4-789b06f2d0d7.png)

As you can see, it's still making 389 requests despite also loading my `build.js` bundle. The bundle is being queued early, but subsequent requests don't stop like they should. Anyone have a guess about what's going on?

**Comment by nlwillia** [comment]
Pretty sure `build.js` in watch mode is intended to replace the regular set of scripts (`system.src.js`, `jspm.browser.js`, `jspm.config.js`, `System.import...`).  Try commenting those out.

**Comment by zakdances** [comment]
@nlwillia But `jspm.browser.js` is the script that references `build.js`. How else does it get injected? Or am I supposed to be editing the script tags in my `index.html` every time I want to use watch mode?

**Comment by kdex** [comment]
I'm experiencing the same issue as @zakdances and experimented a little.

Type | Content | inject flag | System.import("app") triggers module downloads after (!) build/bundle consumption |
| ----- | ------------- | --------- | ----------------------------------------------------------------------- |
| Build | DA  | set | true  |
| Build | DA  | unset  | true  |
| Build | D                           | set | true  |
| Build | D                           | unset   | true |
| Bundle | DA | set | false, but perceivable* latency until loaded |
| Bundle | DA | unset | false, but perceivable* latency until loaded |
| Bundle | D                        | set   | false, but perceivable* latency until loaded |
| Bundle | D                        | unset | false, but perceivable* latency until loaded |

\* I'm measuring 2792 ms in my project, although all modules *should* long be in memory at this time, shouldn't they?
### Legend: 
- *D* = Dependencies only
- *DA* = Dependencies + Application

So obviously, `--inject` doesn't seem to have any effect here, and there seems to be something wrong with the loading time for bundles, too.

So I either have to wait for the dependencies to download for a second time if I create a build, or for an inexplicable (**non-network**) latency for bundles.

I have the feeling that there's at least one bug here. :smile:

**Comment by mgamsjager** [comment]
Chrome network console messes with the download. 

Either use timeline or close the debug tab

**Comment by kdex** [comment]
@mgamsjager I'm still measuring around 1200 ms just to resolve `System.import("app")` when DevTools is completely closed.

**Comment by jackfranklin** [comment]
I might suggest opening a separate issue for this? Purely because this thread is quite long, it's hard to track. Might be easier to debug in separate issue + with a repository that the issue can be reproduced on :)

**Comment by zakdances** [comment]
Yeah, I'll see if I can make a test case for this. Then I'll make a separate issue?

I think this problem is caused by having your source files in a different directory then `jspm.browser.js`. 

My theory: When you bundle, inject, watch, then launch your app in the usual way with something like this:

```
<script>
        SystemJS.import('app.ts');
    </script>
```

SystemJS thinks that `app.ts` is DIFFERENT then the `app.ts` in the bundle because JSPM foolishly assumes everything is it's sibling, and constructs the bundle paths accordingly. And because those paths in `jspm.browser.js` are used as identifiers for the modules, SystemJS doesn't know their supposed to be *replacements* for your source files, not additions.

[Similar issues](https://github.com/jspm/jspm-cli/issues/651) have come up. I've been using the browser-sync `routes` workaround, but obviously that has limitations.

**Comment by zakdances** [comment]
I made a [test case](https://github.com/zakdances/jspmBundleBug) that demonstrates this issue. Let me know if it works.

This issue might also be the same one referenced [here (with more tests available here from @kdex)](https://github.com/jspm/jspm-cli/issues/1707).

**Comment by abhipanda** [comment]
Hello Guys,

I tried the build process and everything is working fine for JS files but for html files the update is not working

Raised an issue for this:

Bundle build not refreshing / updating html files when changed

https://github.com/jspm/jspm-cli/issues/1848

Does anyone know how to solve it  ?

**Comment by jonricaurte** [comment]
@abhipanda @guybedford I think this issue should be closed since there is Watchman integration now which makes reloading bundles almost instant compared to what was used previously for watching bundles.

**Comment by abhipanda** [comment]
It should be.

For the following issue though, we need to add an * at end of the interpreted URL
https://github.com/jspm/jspm-cli/issues/1848

`builder.invalidate(toFileURL(file));`

to

   ` builder.invalidate(toFileURL(file)+'*');`

**Comment by born2net** [comment]
for those of you looking for faster dev times, I recommend hot-loader.
added it to angular2: 
Angular 2 Kitchen sink: http://ng2.javascriptninja.io
and source@ https://github.com/born2net/ng2Boilerplate
Regards,

Sean

**Comment by abhipanda** [comment]
I have been using the new builder in JSPM 17 beta - it works perfectly and I have some 300 files.

**Comment by Munter** [comment]
Hi everyone. I finally got around to finishing the tool I had in mind building for a year or so. I would be really happy if anyone in here, who obviously have dealt with load time performance would give it a go and tell me if it helped them out.

It's an express server middleware that intercepts requests to modules requested by systemjs, then bundles them up with all their dependencies on the fly. In the project we implemented it in we saw massive load time speed ups on initial load and especially in incremental updates. I hope you will see the same performance increase and report any bugs you might run into.

https://github.com/Munter/express-systemjs-translate

**Comment by ArmorDarks** [comment]
I've tested 0.17@beta with `jspm build --watch` flag it seems to be working insanely fast, that's really great. However, I didn't find easy way to hook it up with Browser Sync without launching two parallel processes in console. I guess, it won't be possible, unless jspm's build `--watch` gets API support? Or am I missing something?

**Comment by peteruithoven** [comment]
@ArmorDarks You can easilly combine them in one npm start script, for example:
```
"start": "jspm bundle src/js/index.js -wid & serve & chokidar-socket-emitter",
```

**Comment by ArmorDarks** [comment]
@peteruithoven Thanks for a hint. I've already tried it. Unfortunately, `&` doesn't work on Win machines. `node-run-all` modules, though, solved it and at least it works. Thanks.

I'm just wondering are there any plans to do this in more sane way? While this works, it still looks like a hacking around, especially if you're using Grunt or Gulp. It looks really strange, when part of task runs through Grunt or Gulp, and other part should be run via npm, or hacked around with shell tasks for Grunt or Gulp.

**Comment by oligot** [comment]
Browser sync has some docs on how to use it with [Gulp](https://www.browsersync.io/docs/gulp) and [Grunt](https://www.browsersync.io/docs/grunt).
You could then combine these with the [Jspm API](https://github.com/jspm/jspm-cli/blob/master/docs/api.md) to achieve the same result, but without calling the CLI directly.

The CLI (for both Jspm and Browser sync) seems more appropriate when using npm scripts exclusively, while the Node.js API seems more suited for Gulp/Grunt.

**Comment by oligot** [comment]
Oops, I didn't see that the watch mode is not available in the Jspm API.
There seems to be another way to do it as described in this issue: #1702

**Comment by ArmorDarks** [comment]
@oligot Thanks. Yeah, no API support for watch so far :(

**Comment by pcompassion** [comment]
What if I want to watch multiple app.js at the same time?

Should I run multiple of them?
```
jspm bundle app1.js -wid
jspm bundle app2.js -wid
```

**Comment by Rush** [comment]
Trying different approaches I found best performance with:
* Have express middleware that compiles all javascripts files to ES5/CJS with Babel on the fly and caches the results. Use babel options, and source modified time to generate an `ETag` along with `Cache-Control` set to always with 0 max-age.
* **Do not** use depCache - depCache will add around 400 requests related to loading the babel runtime. 

With all requests delivered with 304 status code the load time with 640 files is around ~3s. Good enough for development. and much faster than watchable bundles.

If anybody is interested in such a middleware I'd be more than welcome to publish it.

**Comment by peteruithoven** [comment]
@Rush I'm definitely interested. So your middleware also watches files? But it only recompiles the modules that changed? 

Did anyone get further with http2/spdy push? [Expush](https://github.com/Munter/expush) lacks documentation and hasn't changed since oct 2015 and it's still an open [issue](https://github.com/mikz/jspm-dev-server/issues/7) on [jspm-dev-server](https://github.com/mikz/jspm-dev-server/).

**Comment by Rush** [comment]
@peteruithoven No need to watch files. There's just one `fs.stat` per request. if the file is modified it will be processed on the fly and cached again, otherwise an already cached version will be served.

**Comment by Rush** [comment]
@peteruithoven I will package my code into a re-usable module tomorrow. One problem I have is inability to easily to synchronize node's babel options with jspm's .. but I do have some ideas on how to improve it.

**Comment by Rush** [comment]
@peteruithoven HTTP2/PUSH is a no-go for development as Chrome and other browsers will refuse to connect to a non-TLS HTTP2 server.

**Comment by peteruithoven** [comment]
> @peteruithoven No need to watch files. There's just one fs.stat per request. if the file is modified (its modify time is newer than the cached version) it will be processed on the fly and cached again, otherwise an already cached version will be served.

That sounds like something that might even save system resources? 

> @peteruithoven HTTP2/PUSH is a no-go for development as Chrome and other browsers will refuse to connect to a non-TLS HTTP2 server (as far as I know).

I know, but you could use a self signed certificate? (I do understand that this is harder to setup)

**Comment by Rush** [comment]
@peteruithoven took longer than anticipated but delivered as promised https://github.com/virtkick/node-serve-jspm

Let me know if it works for you.. I was able to use JSPM's babel and read per file configuration. There was one issue that I was able to workaround https://github.com/jspm/jspm-cli/issues/2202

It requires `jspm@beta` of course

**Comment by Rush** [comment]
@peteruithoven I upgraded serve-jspm to exclude hard-coded babel reference. Now it will run whatever transpiler that's configured in your jspm project.
